core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java

```diff
@@ -198,7 +201,8 @@ private void writeSortedFile(boolean isFinalFile) {
     // spark.shuffle.compress instead of spark.shuffle.spill.compress, so we need to use
     // createTempShuffleBlock here; see SPARK-3426 for more details.
     final Tuple2<TempShuffleBlockId, File> spilledFileInfo =
-      blockManager.diskBlockManager().createTempShuffleBlock();
```

Comment:
one idea: if `isFinalFile` is true, then we call a special version of `createTempShuffleBlock` that takes shuffle & map id, and returns a file path under the same directory of the final shuffle file. Then we don't need to change other places?

Reply:
That would assume the final output has to go to `blockResolver.getDataFile(shuffleId, mapId)` right @cloud-fan  ?
Currently at this layer we do not make that assumption ...

I was initially toying with the idea of passing `mapId` and `shuffleId` as constructor params ... and do something similar when I realized this would make assumptions that the code currently does not make - and so why the base directory is being passed around.

(And then ofcourse I thought we could solve it in `LocalDiskSingleSpillMapOutputWriter` [here](https://github.com/apache/spark/pull/45661#issuecomment-2020527368), but was completely wrong :-( ).